specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,743 @@
+result    ::= “0” | “1” | “2” | [“128” ... ”255”]
+```
+
+The encoding-dependent header may carry metadata or assertions such as the encoded payload length, for integrity and attack proofing purposes. It is not strictly necessary to length-prefix payloads, because req/resp streams are single-use, and stream closures implicitly delimit the boundaries, but certain encodings like SSZ do, for added security.
+
+`encoded-payload` has a maximum byte size of `RQRP_MAX_SIZE`.
+
+Clients MUST ensure the payload size is less than or equal to `RQRP_MAX_SIZE`, if not, they SHOULD reset the stream immediately. Clients tracking peer reputation MAY decrement the score of the misbehaving peer under this circumstance.
+
+#### Requesting side
+
+Once a new stream with the protocol ID for the request type has been negotiated, the full request message should be sent immediately. It should be encoded according to the encoding strategy.
+
+The requester MUST close the write side of the stream once it finishes writing the request message - at this point, the stream will be half-closed.
+
+The requester MUST wait a maximum of **5 seconds** for the first response byte to arrive (time to first byte – or TTFB – timeout). On that happening, the requester will allow further **10 seconds** to receive the full response.
+
+If any of these timeouts fire, the requester SHOULD reset the stream and deem the req/resp operation to have failed.
+
+#### Responding side
+
+Once a new stream with the protocol ID for the request type has been negotiated, the responder must process the incoming request message according to the encoding strategy, until EOF (denoting stream half-closure by the requester).
+
+The responder MUST:
+
+1. Use the encoding strategy to read the optional header.
+2. If there are any length assertions for length `N`, it should read exactly `N` bytes from the stream, at which point an EOF should arise (no more bytes). Should this is not the case, it should be treated as a failure.
+3. Deserialize the expected type, and process the request.
+4. Write the response (result, optional header, payload).
+5. Close their write side of the stream. At this point, the stream will be fully closed.
+
+If steps (1), (2) or (3) fail due to invalid, malformed or inconsistent data, the responder MUST respond in error. Clients tracking peer reputation MAY record such failures, as well as unexpected events, e.g. early stream resets.
+
+The entire request should be read in no more than **5 seconds**. Upon a timeout, the responder SHOULD reset the stream.
+
+The responder SHOULD send a response promptly, starting with a **single-byte** response code which determines the contents of the response (`result` particle in the BNF grammar above).
+
+It can have one of the following values:
+
+-  0: **Success** -- a normal response follows, with contents matching the expected message schema and encoding specified in the request.
```

Comment:
Zero is a default value in some languages, can we set 0 to be UNKNOWN or forbid 0?

Reply:
0 = "no error" - seems appropriate, given that the other codes indicate some form of error - else you have to pick some arbitrary value for "success" in a protocol like this - better go with something that's commonly used this way already. Compare protobuf for example, you'd have an "error" enum in there, you'd set 0 as success because that's what the generated code gives you when the error field lacks a value.

Reply:
Sounds good to me and I agree with that perspective. Thanks!